### PR TITLE
Move token contract source into lib

### PR DESCRIPTION
**Pull request by lexfrl**

### CODE REVIEW

File: src/repo.rs [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+    static LENDER_KEY: H256 = H256([3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
+    static LOAN_TOKEN_KEY: H256 = H256([4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
+    static SECURITY_TOKEN_KEY: H256 = H256([5,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
+    static LOAN_AMOUNT_KEY: H256 = H256([6,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
+    static SECURITY_AMOUNT_KEY: H256 = H256([7,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
+    static INTEREST_RATE_KEY: H256 = H256([8,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
+    static ACTIVATION_DEADLINE_KEY: H256 = H256([9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
+    static RETURN_DEADLINE_KEY: H256 = H256([10,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
+    static BORROW_ACCEPTED_KEY: H256 = H256([11,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
+    static LEND_ACCEPTED_KEY: H256 = H256([12,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]);
```
Comment (NikVolf) [inline src/repo.rs]: could use some macro here to avoid all this zeroes on every line

File: src/repo.rs [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+        }
+        pub fn read_borrower_address(&mut self) -> Address {
+            H256::from(self.storage.read(&BORROWER_KEY)).into()
+        }
+
+        pub fn read_lender_address(&mut self) -> Address {
+            H256::from(self.storage.read(&LENDER_KEY)).into()
+        }
+
+        pub fn read_loan_token_address(&mut self) -> Address {
```
Comment (NikVolf) [inline src/repo.rs]: `read_` along with general rust style to omit `get_` can be omitted as well

File: src/repo.rs [hunk 3]
```diff
@@ -0,0 +1,618 @@
...
+    value: [u8; 32]
+}
+
+struct Storage {
+    table: Vec<Entry>,
+}
+
+/// TODO: optimise by impl hashtable
+impl Storage {
+    fn with_capacity(cap: usize) -> Storage {
```
Comment (NikVolf) [inline src/repo.rs]: why do you need public with_capacity here?
isn't capacity obvious from the usage?

File: src/repo.rs [hunk 4]
```diff
@@ -0,0 +1,618 @@
...
+        fn constructor(&mut self,
+            borrower: Address,
+            lender: Address,
+            loan_token: Address,
+            security_token: Address,
+            loan_amount: U256,
+            security_amount: U256,
+            interest_rate: U256,
+            activation_deadline: u64,
+            return_deadline: u64) {
```
Comment (NikVolf) [inline src/repo.rs]: better use following style for such cases:

```
    activation_deadline: u64,
    return_deadline: u64,
) {
 ...
}
```

File: src/repo.rs [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+            assert!(security_token.transferFrom(borrower_address, this_contract_address, security_amount));
+            // Transfer loan to the borrower address
+            assert!(loan_token.transferFrom(lender_address, borrower_address, loan_amount));
+
+            return true;
+        }
+
+        fn terminate(&mut self) -> bool {
+            if !self.is_active() {
+                panic!("Can't terminate contract: contract hasn't activated");
```
Comment (NikVolf) [inline src/repo.rs]: how to end it if other parity has no desire to activate it?

Comment (lexfrl) [reply src/repo.rs]: anybody can do in through the accept call.. But anyway, I'll add it to terminate too

File: src/repo.rs (hunk 3 above)
Comment (lexfrl) [reply src/repo.rs]: you mean better to just hardcode it?

Comment (NikVolf) [reply src/repo.rs]: yeah, why bothering doing second `with_capacity` function calling the first one despite you use only one value for an argument

File: src/repo.rs [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+                make_token_endpoint(TokenMock::default().with_transfer_from(true)))
+            .build()));
+        let mut contract = default_contract();
+        contract.accept();
+        let spenderExternal = ext_builder().sender(LENDER_ADDR).build();
+        set_external(Box::new(spenderExternal));
+        contract.accept();
+        contract
+    }
+
```
Comment (NikVolf) [inline src/repo.rs]: some redundant empty lines

File: src/repo.rs [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+        fn accept(&mut self) -> bool;
+
+        fn terminate(&mut self) -> bool;
+
+        #[event]
+        fn LendAccepted(&mut self);
+        #[event]
+        fn BorrowAccepted(&mut self);
+
```
Comment (NikVolf) [inline src/repo.rs]: redundant empty line

File: src/repo.rs [hunk 8]
```diff
@@ -0,0 +1,623 @@
...
+    }
+}
+
+pub mod contract {
+    #![allow(non_snake_case)]
+
+    extern crate pwasm_token_contract;
+    use alloc::vec::Vec;
+
+    use pwasm_std::{ext};
```
Comment (NikVolf) [inline src/repo.rs]: why braces?

**Comment by NikVolf** [comment]
lgtm care to move to the dedicated repo?

File: src/repo.rs [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+    }
+
+    fn default_contract() -> RepoContractInstance {
+        let mut contract = RepoContractInstance::new();
+        let loan_amount: U256 = 10000.into();
+        let security_amount: U256 = 50000.into();
+        let interest_rate: U256 = 3.into();
+        let activation_deadline: u64 = 10;
+        let return_deadline: u64 = 20;
+        contract.constructor(BORROWER_ADDR.clone(), LENDER_ADDR.clone(), LOAN_TOKEN_ADDR.clone(), SECURITY_TOKEN_ADDR.clone(),
```
Comment (NikVolf) [inline src/repo.rs]: better to use dedicated line per argument when they all cannot fit one line

**Comment by lexfrl** [comment]
Moved into https://github.com/paritytech/pwasm-repo-contract
I need to merge this request in order to activate CI on pwasm-repo-contract.

**Comment by NikVolf** [comment]
Why json deleted?
Token sample should be *both* binary and lib
Wasm binary and json should be left

**Pull request merged**